### PR TITLE
feat(pci-project): keep project creation state on reload

### PR DESCRIPTION
**Pull request by FabienHenon**
ref: #MANAGER-19526

## Description

Ticket Reference: #...

### CODE REVIEW

File: packages/manager/apps/pci-project/src/components/payment/RegisterPaymentMethod.tsx [hunk 1]
```diff
@@ -77,6 +75,25 @@ const RegisterPaymentMethod: React.FC<RegisterPaymentMethodProps> = ({
...
+      const method = availablePaymentMethods.find(
+        (m) => m.paymentType === preselectedPaymentType,
+      );
+      if (method) {
+        onHandlePaymentMethodChange(method);
+      }
+    }
+  }, [preselectedPaymentType, selectedPaymentMethod, availablePaymentMethods]);
+
+  if (isLoading || isFFLoading || !availablePaymentMethods || !features) {
```
Comment (StephaneRavet) [inline packages/manager/apps/pci-project/src/components/payment/RegisterPaymentMethod.tsx:98]: ⚠️ **Loading state issue**: The component returns a spinner before the useEffect hook runs, which could prevent the preselection logic from executing properly. Consider moving the useEffect before the loading check or adding a flag to track if preselection has been attempted.

File: packages/manager/apps/pci-project/src/components/payment/RegisterPaymentMethod.tsx [hunk 2]
```diff
@@ -77,6 +75,25 @@ const RegisterPaymentMethod: React.FC<RegisterPaymentMethodProps> = ({
...
+      !selectedPaymentMethod
+    ) {
+      const method = availablePaymentMethods.find(
+        (m) => m.paymentType === preselectedPaymentType,
+      );
+      if (method) {
+        onHandlePaymentMethodChange(method);
+      }
+    }
+  }, [preselectedPaymentType, selectedPaymentMethod, availablePaymentMethods]);
```
Comment (StephaneRavet) [inline packages/manager/apps/pci-project/src/components/payment/RegisterPaymentMethod.tsx]: ⚠️ **Missing dependency**: The useEffect dependency array is missing `onHandlePaymentMethodChange`. This could cause stale closure issues. Consider using useCallback for the handler or add it to the dependencies.

File: packages/manager/apps/pci-project/src/pages/creation/Creation.page.tsx [hunk 3]
```diff
@@ -42,8 +45,36 @@ export default function ProjectCreation() {
...
+  const cart = hasInitialCart ? existingCart : createdCart;
+
+  const {
+    mutate: orderProjectItem,
+    data: createdProjectItem,
+  } = useOrderProjectItem();
+  const { data: existingProjectItem } = useGetOrderProjectId(
+    hasInitialCart ? searchParams.get('cartId') || undefined : undefined,
+  );
+
```
Comment (StephaneRavet) [inline packages/manager/apps/pci-project/src/pages/creation/Creation.page.tsx:75]: 🔧 **Complex state logic**: The Creation page has grown quite complex with multiple state management paths (hasInitialCart, existing vs created items). Consider extracting this logic into a custom hook like `useCartState` to improve readability and testability.

File: packages/manager/apps/pci-project/src/pages/creation/Creation.page.tsx [hunk 4]
```diff
@@ -60,17 +91,36 @@ export default function ProjectCreation() {
...
-        onSuccess: (createdCart) => {
+        onSuccess: (justCreatedCart) => {
           orderProjectItem({
-            cartId: createdCart.cartId,
+            cartId: justCreatedCart.cartId,
           });
+          setSearchParams({ cartId: justCreatedCart.cartId });
         },
       },
     );
```
Comment (StephaneRavet) [inline packages/manager/apps/pci-project/src/pages/creation/Creation.page.tsx:122]: ⚠️ **Missing error handling**: The useEffect hook that creates the cart doesn't handle potential errors. If cart creation fails, the user will be stuck without feedback. Consider adding error handling and user feedback.

File: packages/manager/apps/pci-project/src/components/payment/RegisterPaymentMethod.spec.tsx [hunk 5]
```diff
@@ -506,6 +506,107 @@ describe('RegisterPaymentMethod', () => {
...
+          {...defaultProps}
+          preselectedPaymentType={TPaymentMethodType.PAYPAL}
+          handlePaymentMethodChange={handlePaymentMethodChange}
+        />,
+      );
+
+      expect(handlePaymentMethodChange).toHaveBeenCalledWith(
+        mockAvailablePaymentMethods[1],
+      );
+    });
```
Comment (StephaneRavet) [inline packages/manager/apps/pci-project/src/components/payment/RegisterPaymentMethod.spec.tsx:607]: ✅ **Good testing**: Excellent test coverage for the preselection functionality, including edge cases like waiting for data to load and preventing multiple preselections.

File: packages/manager/apps/pci-project/src/pages/creation/Creation.page.tsx (hunk 4 above)
Comment (FabienHenon) [reply packages/manager/apps/pci-project/src/pages/creation/Creation.page.tsx:122]: Errors are handled by the ErrorBoundary.
Also, this component is not finished yet. The goal of this PR was only to add state recovery

File: packages/manager/apps/pci-project/src/pages/creation/Creation.page.tsx (hunk 3 above)
Comment (FabienHenon) [reply packages/manager/apps/pci-project/src/pages/creation/Creation.page.tsx:75]: That's something I plan on doing as soon as we finish the project creation process. Doing it now will require more work as we will have to do it again later when the project creation component is finished

File: packages/manager/apps/pci-project/src/components/payment/RegisterPaymentMethod.tsx (hunk 1 above)
Comment (FabienHenon) [reply packages/manager/apps/pci-project/src/components/payment/RegisterPaymentMethod.tsx:98]: I don't think it prevents the hooks from running and calling the necessary callback

File: packages/manager/apps/pci-project/src/pages/creation/Creation.page.tsx [hunk 6]
```diff
@@ -42,8 +45,36 @@ export default function ProjectCreation() {
...
 
-  const { mutate: createAndAssignCart, data: cart } = useCreateAndAssignCart();
-  const { mutate: orderProjectItem, data: projectItem } = useOrderProjectItem();
+  const {
+    mutate: createAndAssignCart,
+    data: createdCart,
+  } = useCreateAndAssignCart();
+  const { data: existingCart } = useGetCart(
+    hasInitialCart ? searchParams.get('cartId') || undefined : undefined,
+  );
```
Comment (seven-amid) [inline packages/manager/apps/pci-project/src/pages/creation/Creation.page.tsx:54]: `  const { data: existingCart } = useGetCart(searchParams.get('cartId'));
`
Think u can write it this way, since the getCart is enabled only when cartId is defined

File: packages/manager/apps/pci-project/src/pages/creation/Creation.page.tsx [hunk 7]
```diff
@@ -42,8 +45,36 @@ export default function ProjectCreation() {
...
+
+  const {
+    form,
+    setForm,
+    isConfigFormValid,
+    isLoading: isLoadingConfigForm,
+  } = useConfigForm(
+    ovhSubsidiary,
+    hasInitialCart ? searchParams.get('cartId') || undefined : undefined,
```
Comment (seven-amid) [inline packages/manager/apps/pci-project/src/pages/creation/Creation.page.tsx:63]: same remark here

File: packages/manager/apps/pci-project/src/pages/creation/Creation.page.tsx [hunk 8]
```diff
@@ -60,17 +91,36 @@ export default function ProjectCreation() {
   } = useStepper();
 
   useEffect(() => {
+    if (isLoadingConfigForm) {
+      return;
+    }
+
+    if (!isConfigFormValid()) {
+      return;
+    }
```
Comment (seven-amid) [inline packages/manager/apps/pci-project/src/pages/creation/Creation.page.tsx:100]:
```
if(isLoadingConfigForm || !isConfigFormValid()) {
 return;
}
```

**Review by seven-amid — APPROVED** [review]

File: packages/manager/apps/pci-project/src/pages/creation/Creation.page.tsx (hunk 6 above)
Comment (FabienHenon) [reply packages/manager/apps/pci-project/src/pages/creation/Creation.page.tsx:54]: Yes but after cart creation its id is added to the query params, thus `searchParams.get('cartId')` becomes non undefined and the cart is retrieved (which is not what we want here)

File: packages/manager/apps/pci-project/src/pages/creation/Creation.page.tsx (hunk 7 above)
Comment (FabienHenon) [reply packages/manager/apps/pci-project/src/pages/creation/Creation.page.tsx:63]: Same answer ;)

File: packages/manager/apps/pci-project/src/pages/creation/Creation.page.tsx (hunk 8 above)
Comment (FabienHenon) [reply packages/manager/apps/pci-project/src/pages/creation/Creation.page.tsx:100]: We must not check if the config form is valid if it's loading. That's why I separated the ifs.
With your solution, I think, indeed that if the first part is true it won't execute the second part, but can we rely on this?
